File: 11-20/BM20-数组中的逆序对/main.cpp
#include <bits/stdc++.h>
using namespace std;

#define FOR_INC(i, start, end) for (int i = start; i < end; ++i)
#define FOR_DEC(i, start, end) for (int i = start; i > end; --i)
#define FOR_INC_EQUAL(i, start, end) for (int i = start; i <= end; ++i)
#define FOR_DEC_EQUAL(i, start, end) for (int i = start; i >= end; --i)

class Solution {
public:
    int ans = 0;
    const int mod = int(pow(10,9))+7;
    void merge_sort(vector<int>& nums,int left,int right,vector<int>& tmp){
        if(left==right){
            return;
        }
        int mid = left+(right-left)/2;
        int i = left , j = mid+1 ,index = left;
        merge_sort(nums,left,mid,tmp);
        merge_sort(nums,mid+1,right,tmp);
        while(i<=mid&&j<=right){
            if(nums[i]<nums[j]){
                tmp[index++]=nums[i++];
            }else{ //后面大于前面，是逆序对
                tmp[index++]=nums[j++];
                ans+=mid-i+1;//逆序对个数mid-i+1
                ans%=mod;
            }
        }
        while(i<=mid){
            tmp[index++]=nums[i++];
        }
        while(j<=right){
            tmp[index++]=nums[j++];
        }
        FOR_INC_EQUAL(p,left,right){
            nums[p] = tmp[p];
        }
    }
    int InversePairs(vector<int>& nums) {//归并排序
        // write code here
        int n = nums.size();
        vector<int> tmp(n,0);
        merge_sort(nums,0,n-1,tmp);
        cout<<this->ans<<endl;
        return this->ans;
    }
};
int main()
{
    Solution s;
    vector<int> nums{1,2,3,4,5,6,7,0};
    s.InversePairs(nums);
    return 0;
}

File: 11-20/BM20-数组中的逆序对/main.java
import java.util.*;
public class Main {

    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    public static class ListNode {
        int val;
        ListNode next = null;
        public ListNode(int val) {
            this.val = val;
        }
    }

    public void getNums(String s){
        StringBuffer sb = new StringBuffer();
        for(int i=0;i<s.length();i++){
            char ch = s.charAt(i);
            if(ch=='['){
                sb.append('{');
            }else if(ch==']'){
                sb.append('}');
            }else if(ch=='"'){
                sb.append("'");
            } else{
                sb.append(ch);
            }
        }
        System.out.println(sb.toString());
    }


    public int InversePairs (int[] nums) { //二分查找
        // write code here
        final int mod = 1000000007;
        int ans = 0;
        List<Integer> list = new ArrayList<>();
        int n = nums.length;
        for(int i=n-1;i>=0;i--){
            int val = nums[i];
            if(list.isEmpty()){
                list.add(val);
            }
            else{
                int left = 0;
                int right = list.size()-1;
                int mid;
                while (left<=right){//寻找插入位置
                    mid = left+(right-left)/2;
                    int temp = list.get(mid);
                    if(temp<val){
                        left=mid+1;
                    }
                    else{
                        right = mid-1;
                    }
                }
                ans+=left;//有几个数比它大
                ans%=mod;
                list.add(left,val);//插入指定位置
            }
        }
        return ans;
    }
    
    public static void main(String[] args) {
        Main main = new Main();
    }
}
